In [78]:
import pandas as pd

# Specify the file path
file_path = "../../data/translated_comments.xlsx"

# Load the Excel file into a DataFrame
df = pd.read_excel(file_path)

In [79]:
# Load a specific sheet named 'Sheet1'
df = pd.read_excel(file_path, sheet_name='Sheet1')


In [80]:
df.head()

,Comment Id,Comment
0,7221188022045099008,rashmika වාසිය
1,7221178613064123392,rashmika වෙගී ade
2,7221176459355882496,සාරංග
3,7333542921793094656,ඉස්සරහට නිලියක් ශුවර්
4,7221231966270751744,rashmika මතකේ una..eya වාසිය ටිකක


In [81]:
# Get the row count
row_count = df.shape[0]

print("Row count of the DataFrame:", row_count)

Row count of the DataFrame: 14017


In [5]:
# Find null values in the "Comment" column
null_comments = df[df['Comment'].isnull()]

# Display null values in the "Comment" column
print("Null values in the 'Comment' column:")
print(null_comments)

Null values in the 'Comment' column:
                Comment Id Comment
44     7221177654576825344     NaN
45     7221176223815713792     NaN
48     7243025175670604800     NaN
50     7224778042249839616     NaN
52     7223030238438032384     NaN
...                    ...     ...
13813  7304609734580945920     NaN
13814  7301703176813511680     NaN
13815  7294979703093068800     NaN
13816  7294296586959471616     NaN
13890  7292693117953934336     NaN

[3122 rows x 2 columns]


In [6]:
# Drop rows with null values in the "Comment" column
df = df.dropna(subset=['Comment'])

In [7]:
# Get the row count
row_count = df.shape[0]

print("Row count of the DataFrame:", row_count)

Row count of the DataFrame: 10895


In [8]:
# Specify the file path for the new file
output_file_path = "../../data/finale_dataset_v1.xlsx"

# Save the DataFrame to a new Excel file
df.to_excel(output_file_path, index=False)

print("DataFrame saved to", output_file_path)

DataFrame saved to ../../data/finale_dataset_v1.xlsx


## Starting the BERT model fine tune for the dataset

In [9]:
pip install transformers datasets

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd

# Specify the file path
file_path = "../../data/mnt/sentiment_Aspect_added.xlsx"

# Load the Excel file into a DataFrame
df = pd.read_excel(file_path)


In [3]:
df.head()

,Comment ID,Comment,Business Sentiment,Content Sentiment
0,7.220000e+18,rashmika වගේ,Positive,Positive
1,7.220000e+18,rashmika වෙගේ අඩේ,Positive,Positive
2,7.220000e+18,සාරංග,Neutral,Neutral
3,7.330000e+18,ඉස්සරහට නිලියක් ශුවර්,Positive,Positive
4,7.220000e+18,rashmika මතක් උනා. එයා වගේ ටිකක්,Positive,Positive


In [4]:
import pandas as pd
from sklearn.model_selection import train_test_split
from datasets import Dataset

# Load the dataset
file_path = '../../data/mnt/sentiment_Aspect_added.xlsx'
data = pd.read_excel(file_path)

# Split the dataset into training and testing sets
train_data, test_data = train_test_split(data, test_size=0.2, random_state=42)

# Convert pandas dataframes to Hugging Face datasets
train_dataset = Dataset.from_pandas(train_data)
test_dataset = Dataset.from_pandas(test_data)

# Verify the split
print(f"Training set size: {len(train_dataset)}")
print(f"Testing set size: {len(test_dataset)}")

Training set size: 568
Testing set size: 142


In [5]:
# Ensure the column name is correct
assert 'Comment' in data.columns, "Column 'Comment' not found in the dataset."

In [6]:
data.head()

,Comment ID,Comment,Business Sentiment,Content Sentiment
0,7.220000e+18,rashmika වගේ,Positive,Positive
1,7.220000e+18,rashmika වෙගේ අඩේ,Positive,Positive
2,7.220000e+18,සාරංග,Neutral,Neutral
3,7.330000e+18,ඉස්සරහට නිලියක් ශුවර්,Positive,Positive
4,7.220000e+18,rashmika මතක් උනා. එයා වගේ ටිකක්,Positive,Positive


In [7]:
data['Comment'].head()

0                        rashmika වගේ
1                   rashmika වෙගේ අඩේ
2                               සාරංග
3               ඉස්සරහට නිලියක් ශුවර්
4    rashmika මතක් උනා. එයා වගේ ටිකක්
Name: Comment, dtype: object

In [8]:
comments = data['Comment'].tolist()  # Ensure it's a list of strings
for comment in comments:
    print(comment)

rashmika වගේ
rashmika වෙගේ අඩේ
සාරංග
ඉස්සරහට නිලියක් ශුවර්
rashmika මතක් උනා. එයා වගේ ටිකක්
ව්‍යාකූල උනා පිපිරීම මීට සාරන්ග
මේ අර සිකුරු ඇවිත් ට්‍රාමා එකේ අන්ජලී නේද?
එයා නම් ලස්සනයි
She is ලස්සන
rashmika මඳන වාගී
අනවශ්‍ය තොරතුරක්...
සාරංග වම
ආහ... අනවශ්‍ය තොරතුරක් ..
සාරංගයා
සාරංග ලෙෆ්ට්
වාව් රශ්මිකා වගේ
nan
@
මෙ වගෙ නිලියක් ඉන්නො නෙඩ
සාරාංග
@Navee කෙල්ල මේ අර එදාඅ කිවුවේ
ප් රතික් රියාව එක උගේ
සාරංග ව් යාකූලත්වය
සාරංග මරැ
Hayyo සාරංග
rashmika වාසිය
rashmika වාසිය
adeh සිරාවටම අහේ awlakda rashmika වාගෙනෙ
@Dinakshie Saranga Dina හොයමින් සුන්දර
She's සුන්දර
අම්මොහ් සාරංග
just හරියට mandana....
rashmikage හිනාව මයි
ව් යාකූලත්වය උනා සාරංග
සාරංග
එයාගේ තව නැද්ද
rashmika මඳන vage...
rashmika වාසිය
සාරංග අයියා
Tikak Rashmika වාගී
බලන්න හරියට a rashmika @iamrashmika
බලන්න හරියට rashmika
The few momments පස්සෙ
rashmika
rasmika wagemi
රශ්මිකා වගේ එයා
දෙරණ ඩ්‍රාමා එකක මෙයා ඇක්ට් කරනවා නේද
ලස්සන
@anjaliwathsalajayasinghe
@Dinakshie Saranga අයියගේ මූණ
ව් යාකූලත්වය
සාරංග අයියා
Anjale අක්ක
tr@Trading 

### load the SinhalaBERTo and created preproccess funtion 

In [9]:
import pandas as pd
from transformers import AutoModelForSequenceClassification, AutoTokenizer
import torch

In [10]:
data.head()

,Comment ID,Comment,Business Sentiment,Content Sentiment
0,7.220000e+18,rashmika වගේ,Positive,Positive
1,7.220000e+18,rashmika වෙගේ අඩේ,Positive,Positive
2,7.220000e+18,සාරංග,Neutral,Neutral
3,7.330000e+18,ඉස්සරහට නිලියක් ශුවර්,Positive,Positive
4,7.220000e+18,rashmika මතක් උනා. එයා වගේ ටිකක්,Positive,Positive


In [11]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("keshan/SinhalaBERTo")

In [12]:
tokenizer

RobertaTokenizerFast(name_or_path='keshan/SinhalaBERTo', vocab_size=52000, model_max_length=1000000000000000019884624838656, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'bos_token': '<s>', 'eos_token': '</s>', 'unk_token': '<unk>', 'sep_token': '</s>', 'pad_token': '<pad>', 'cls_token': '<s>', 'mask_token': '<mask>'}, clean_up_tokenization_spaces=True),  added_tokens_decoder={
	0: AddedToken("<s>", rstrip=False, lstrip=False, single_word=False, normalized=True, special=True),
	1: AddedToken("<pad>", rstrip=False, lstrip=False, single_word=False, normalized=True, special=True),
	2: AddedToken("</s>", rstrip=False, lstrip=False, single_word=False, normalized=True, special=True),
	3: AddedToken("<unk>", rstrip=False, lstrip=False, single_word=False, normalized=True, special=True),
	4: AddedToken("<mask>", rstrip=False, lstrip=True, single_word=False, normalized=False, special=True),
}

### Analyse the tokenizer

In [13]:
# Vocabulary size
vocab_size = tokenizer.vocab_size
print(f"Vocabulary Size: {vocab_size}")

Vocabulary Size: 52000


In [14]:
# Tokenization example
text = "rashmika වගේ"
tokens = tokenizer.tokenize(text)
print(f"Tokens: {tokens}")

Tokens: ['r', 'ash', 'm', 'ika', 'Ġà·Ģà¶ľ', 'à·ļ']


In [15]:
# Token IDs
token_ids = tokenizer.convert_tokens_to_ids(tokens)
print(f"Token IDs: {token_ids}")

Token IDs: [86, 2698, 81, 4920, 415, 277]


In [16]:
# Decoding token IDs back to text
decoded_text = tokenizer.decode(token_ids)
print(f"Decoded Text: {decoded_text}")

Decoded Text: rashmika වගේ


In [17]:
# Special tokens
special_tokens = tokenizer.special_tokens_map
print(f"Special Tokens: {special_tokens}")

Special Tokens: {'bos_token': '<s>', 'eos_token': '</s>', 'unk_token': '<unk>', 'sep_token': '</s>', 'pad_token': '<pad>', 'cls_token': '<s>', 'mask_token': '<mask>'}


In [18]:
# Padding and truncation
encoded_input = tokenizer(text, padding='max_length', truncation=True, max_length=10)
print(f"Encoded Input with Padding and Truncation: {encoded_input}")

Encoded Input with Padding and Truncation: {'input_ids': [0, 86, 2698, 81, 4920, 415, 277, 2, 1, 1], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 0, 0]}


#### [0, 86, 2698, 81, 4920, 415, 277, 2, 1, 1]
0 is typically the ID for the [CLS] token, which is added at the beginning of the input sequence to signify the start of the sequence.
2 is typically the ID for the [SEP] token, which is added at the end of the input sequence to signify the end of the sequence.
1, 1 are padding tokens added to make the total length 10. These are often represented by the ID 1.

#### [1, 1, 1, 1, 1, 1, 1, 1, 0, 0]
This mask is used to distinguish between real tokens and padding tokens.
1 indicates that the corresponding token in input_ids is a real token.
0 indicates that the corresponding token in input_ids is a padding token.  

In [19]:
# Encoding example
encoded = tokenizer.encode(text)
print(f"Encoded Text: {encoded}")

Encoded Text: [0, 86, 2698, 81, 4920, 415, 277, 2]


In [20]:
from transformers import AutoModelForSequenceClassification, AutoTokenizer

model = AutoModelForSequenceClassification.from_pretrained("keshan/SinhalaBERTo")
tokenizer = AutoTokenizer.from_pretrained("keshan/SinhalaBERTo")

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at keshan/SinhalaBERTo and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


### Define a function to predict sentiment

In [21]:
# def predict_sentiment(comment):
#     inputs = tokenizer(comment, return_tensors="pt", truncation=True, padding=True)
#     with torch.no_grad():
#         outputs = model(**inputs)
#     logits = outputs.logits
#     predicted_class_id = logits.argmax().item()
#     return model.config.id2label[predicted_class_id]

In [22]:
# # Ensure each comment is passed as a string when applying the function
# data['Predicted Sentiment'] = data['Comment'].astype(str).apply(predict_sentiment)

In [23]:
data.head()

,Comment ID,Comment,Business Sentiment,Content Sentiment
0,7.220000e+18,rashmika වගේ,Positive,Positive
1,7.220000e+18,rashmika වෙගේ අඩේ,Positive,Positive
2,7.220000e+18,සාරංග,Neutral,Neutral
3,7.330000e+18,ඉස්සරහට නිලියක් ශුවර්,Positive,Positive
4,7.220000e+18,rashmika මතක් උනා. එයා වගේ ටිකක්,Positive,Positive


In [24]:
# # Compare the predicted sentiment with existing sentiment columns
# data['Business Match'] = data['Business Sentiment'] == data['Predicted Sentiment']
# data['Content Match'] = data['Content Sentiment'] == data['Predicted Sentiment']

In [25]:
# # Print the updated dataset
# data.head(20)

In [26]:
# inspect the label mapping

label_mappings = model.config.id2label
print(label_mappings)

{0: 'LABEL_0', 1: 'LABEL_1'}


In [27]:
def predict_sentiment(comment):
    inputs = tokenizer(comment, return_tensors="pt", truncation=True, padding=True)
    with torch.no_grad():
        outputs = model(**inputs)
    logits = outputs.logits
    predicted_class_id = logits.argmax().item()
    return predicted_class_id

# positive_sentences = ["ඒක හොඳයි.", "මට හරි සතුටුයි."]
# negative_sentences = ["ඒක නරකයි.", "මට සතුටු නෑ"]

positive_sentences = ["ශුවර්"]
negative_sentences = ["ඒක සුපිරි නෑ.", "ලස්සනයි ඒක"]


for sentence in positive_sentences:
    print(f"Positive sentence: '{sentence}' -> Label: {predict_sentiment(sentence)}")
for sentence in negative_sentences:
    print(f"Negative sentence: '{sentence}' -> Label: {predict_sentiment(sentence)}")

Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


Positive sentence: 'ශුවර්' -> Label: 0
Negative sentence: 'ඒක සුපිරි නෑ.' -> Label: 0
Negative sentence: 'ලස්සනයි ඒක' -> Label: 0


In [28]:
positive_sentences = ["පුදුමයි.", "ආදරෙයි."]
neutral_sentences = ["මාර වැඩක්.", "එල."]
negative_sentences = ["I hate it.", "This is terrible."]

for sentence in positive_sentences:
    print(f"Positive: '{sentence}' -> Label: {predict_sentiment(sentence)}")

for sentence in neutral_sentences:
    print(f"Neutral: '{sentence}' -> Label: {predict_sentiment(sentence)}")

for sentence in negative_sentences:
    print(f"Negative: '{sentence}' -> Label: {predict_sentiment(sentence)}")

Positive: 'පුදුමයි.' -> Label: 0
Positive: 'ආදරෙයි.' -> Label: 0
Neutral: 'මාර වැඩක්.' -> Label: 0
Neutral: 'එල.' -> Label: 0
Negative: 'I hate it.' -> Label: 0
Negative: 'This is terrible.' -> Label: 0


## Fine tune the model with custom data

In [29]:
pip install transformers datasets torch

Note: you may need to restart the kernel to use updated packages.


In [30]:
data.head()

,Comment ID,Comment,Business Sentiment,Content Sentiment
0,7.220000e+18,rashmika වගේ,Positive,Positive
1,7.220000e+18,rashmika වෙගේ අඩේ,Positive,Positive
2,7.220000e+18,සාරංග,Neutral,Neutral
3,7.330000e+18,ඉස්සරහට නිලියක් ශුවර්,Positive,Positive
4,7.220000e+18,rashmika මතක් උනා. එයා වගේ ටිකක්,Positive,Positive


In [31]:
import pandas as pd

# Specify the file path
file_path = "../../data/mnt/sentiment_Aspect_added.xlsx"

# Load the Excel file into a DataFrame
df = pd.read_excel(file_path)


In [32]:
df.head()

,Comment ID,Comment,Business Sentiment,Content Sentiment
0,7.220000e+18,rashmika වගේ,Positive,Positive
1,7.220000e+18,rashmika වෙගේ අඩේ,Positive,Positive
2,7.220000e+18,සාරංග,Neutral,Neutral
3,7.330000e+18,ඉස්සරහට නිලියක් ශුවර්,Positive,Positive
4,7.220000e+18,rashmika මතක් උනා. එයා වගේ ටිකක්,Positive,Positive


In [33]:
# Get unique sentiment values in 'Business Sentiment' and 'Content Sentiment'
unique_business_sentiments = data['Business Sentiment'].unique()
unique_content_sentiments = data['Content Sentiment'].unique()

print("Unique Business Sentiments:", unique_business_sentiments)
print("Unique Content Sentiments:", unique_content_sentiments)

Unique Business Sentiments: ['Positive' 'Neutral' 'Negative' nan]
Unique Content Sentiments: ['Positive' 'Neutral' 'Negative' nan 'Negative (Sarcastic criticism)'
 'Negative (could be sarcastic)' 'Negative (Could be sarcastic)'
 'Negative (Comparison)']


In [34]:
# Count the number of NaN values in 'Business Sentiment' and 'Content Sentiment'
nan_business_sentiment = data['Business Sentiment'].isna().sum()
nan_content_sentiment = data['Content Sentiment'].isna().sum()

print("Number of NaN values in Business Sentiment:", nan_business_sentiment)
print("Number of NaN values in Content Sentiment:", nan_content_sentiment)

Number of NaN values in Business Sentiment: 1
Number of NaN values in Content Sentiment: 1


In [35]:


# Sample DataFrame with equal lengths for both columns
data = pd.DataFrame({
    'Business Sentiment': ['Positive', 'Neutral', 'Negative', 'Positive', 'Neutral', 'Negative', 'Neutral', 'Positive'],
    'Content Sentiment': ['Positive', 'Neutral', 'Negative', 'Negative (Sarcastic criticism)', 
                          'Negative (could be sarcastic)', 'Negative (Could be sarcastic)', 
                          'Negative (Comparison)', 'Positive']
})

# Replace specific "Negative" related sentiments with "Negative"
negative_related_sentiments = ['Negative (Sarcastic criticism)', 
                               'Negative (could be sarcastic)', 
                               'Negative (Could be sarcastic)', 
                               'Negative (Comparison)']

data['Content Sentiment'] = data['Content Sentiment'].replace(negative_related_sentiments, 'Negative')

# Drop rows with NaN values in 'Business Sentiment' and 'Content Sentiment'
data = data.dropna(subset=['Business Sentiment', 'Content Sentiment'])

In [36]:
unique_business_sentiments = data['Business Sentiment'].unique()
unique_content_sentiments = data['Content Sentiment'].unique()

print("Unique Business Sentiments:", unique_business_sentiments)
print("Unique Content Sentiments:", unique_content_sentiments)

Unique Business Sentiments: ['Positive' 'Neutral' 'Negative']
Unique Content Sentiments: ['Positive' 'Neutral' 'Negative']


In [37]:
data.head(5)

,Business Sentiment,Content Sentiment
0,Positive,Positive
1,Neutral,Neutral
2,Negative,Negative
3,Positive,Negative
4,Neutral,Negative


In [38]:
len(data)

8

In [39]:
import pandas as pd
from sklearn.model_selection import train_test_split
from datasets import Dataset

# Load the dataset
file_path = '../../data/mnt/sentiment_Aspect_added.xlsx'
data = pd.read_excel(file_path)

# Ensure all comments are strings
data['Comment'] = data['Comment'].astype(str)

# Split the dataset into training and testing sets
train_data, test_data = train_test_split(data, test_size=0.2, random_state=42)

# Convert pandas dataframes to Hugging Face datasets
train_dataset = Dataset.from_pandas(train_data)
test_dataset = Dataset.from_pandas(test_data)

# Drop the __index_level_0__ column
train_dataset = train_dataset.remove_columns(['__index_level_0__'])
test_dataset = test_dataset.remove_columns(['__index_level_0__'])

# Verify the split
print(f"Training set size: {len(train_dataset)}")
print(f"Testing set size: {len(test_dataset)}")


Training set size: 568
Testing set size: 142


In [40]:
data.head()

,Comment ID,Comment,Business Sentiment,Content Sentiment
0,7.220000e+18,rashmika වගේ,Positive,Positive
1,7.220000e+18,rashmika වෙගේ අඩේ,Positive,Positive
2,7.220000e+18,සාරංග,Neutral,Neutral
3,7.330000e+18,ඉස්සරහට නිලියක් ශුවර්,Positive,Positive
4,7.220000e+18,rashmika මතක් උනා. එයා වගේ ටිකක්,Positive,Positive


 The column __index_level_0__ was likely added when you converted the pandas DataFrame to Hugging Face datasets using the Dataset.from_pandas method. 

In [41]:
train_dataset

Dataset({
    features: ['Comment ID', 'Comment', 'Business Sentiment', 'Content Sentiment'],
    num_rows: 568
})

In [42]:
test_dataset

Dataset({
    features: ['Comment ID', 'Comment', 'Business Sentiment', 'Content Sentiment'],
    num_rows: 142
})

In [43]:
# Print labels for checking
print(f"Train_Dataset: {train_dataset}")
print(f"Ltrain_dataset types: {[type(train_dataset) for train_dataset in train_dataset[0]]}")

Train_Dataset: Dataset({
    features: ['Comment ID', 'Comment', 'Business Sentiment', 'Content Sentiment'],
    num_rows: 568
})
Ltrain_dataset types: [<class 'str'>, <class 'str'>, <class 'str'>, <class 'str'>]


the train_dataset and test_dataset are not pandas dataframes. They are instances of the datasets.Dataset class from the Hugging Face datasets library. This class is specifically designed for handling and processing large datasets used in machine learning and natural language processing tasks.

In [44]:
# Print the column names
print("Columns in train_dataset:", train_dataset.column_names)

# Iterate through the dataset and print the values for each column
for column in train_dataset.column_names:
    print(f"\nColumn: {column}")
    print(train_dataset[column][:5])  # Display the first 5 values for brevity

Columns in train_dataset: ['Comment ID', 'Comment', 'Business Sentiment', 'Content Sentiment']

Column: Comment ID
[7.35e+18, 7.34e+18, 7.22e+18, 7.21e+18, 7.35e+18]

Column: Comment
['Woah', '@Henuka Vithanage', 'Hayyo සාරංග', 'මම පාවිච්චි කරන්නෙත් iphone එකක් මචන්', 'maroons']

Column: Business Sentiment
['Positive', 'Neutral', 'Negative', 'Neutral', 'Neutral']

Column: Content Sentiment
['Positive', 'Neutral', 'Negative', 'Neutral', 'Neutral']


In [45]:
# # Convert to pandas DataFrame
# train_df = train_dataset.to_pandas()

# # Display the DataFrame
# print(train_df.head())  # View the first few rows

 identify all the unique sentiment values in both the "Business Sentiment" and "Content Sentiment" columns.

In [46]:
# Replace specific "Negative" related sentiments with "Negative"
negative_related_sentiments = ['Negative (Sarcastic criticism)', 
                               'Negative (could be sarcastic)', 
                               'Negative (Could be sarcastic)', 
                               'Negative (Comparison)',
                              'Neutral']

data['Content Sentiment'] = data['Content Sentiment'].replace(negative_related_sentiments, 'Negative')
data['Business Sentiment'] = data['Business Sentiment'].replace(negative_related_sentiments, 'Negative')

# Drop rows with NaN values in 'Business Sentiment' and 'Content Sentiment'
data = data.dropna(subset=['Business Sentiment', 'Content Sentiment'])

# Verify that only the expected sentiments are present
unique_business_sentiments = data['Business Sentiment'].unique()
unique_content_sentiments = data['Content Sentiment'].unique()

print("Unique Business Sentiments:", unique_business_sentiments)
print("Unique Content Sentiments:", unique_content_sentiments)

Unique Business Sentiments: ['Positive' 'Negative']
Unique Content Sentiments: ['Positive' 'Negative']


In [47]:
# Load the tokenizer and model
tokenizer = AutoTokenizer.from_pretrained("keshan/SinhalaBERTo")
model = AutoModelForSequenceClassification.from_pretrained("keshan/SinhalaBERTo", num_labels=2)  # Adjust num_labels if necessary

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at keshan/SinhalaBERTo and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [75]:
from sklearn.model_selection import train_test_split
import numpy as np
from transformers import AutoTokenizer, AutoModelForSequenceClassification
from datasets import Dataset

sentiment_mapping = {
    "Positive": 1,
    "Negative": 0
}

# Define the preprocessing function for Business Sentiment
def preprocess_business(examples):
    comments = examples['Comment']
    business_sentiment = examples['Business Sentiment']
    
    # Tokenize the inputs
    tokenized_inputs = tokenizer(comments, padding="max_length", truncation=True, max_length=128)
    
    # Convert business sentiment labels to numerical values
    labels = [sentiment_mapping[business] for business in business_sentiment]
    
    # Convert labels to a numpy array for Trainer compatibility
    tokenized_inputs["labels"] = np.array(labels)
    
    return tokenized_inputs

# Define the preprocessing function for Content Sentiment
def preprocess_content(examples):
    comments = examples['Comment']
    content_sentiment = examples['Content Sentiment']
    
    # Tokenize the inputs
    tokenized_inputs = tokenizer(comments, padding="max_length", truncation=True, max_length=128)
    
    # Convert content sentiment labels to numerical values
    labels = [sentiment_mapping[content] for content in content_sentiment]
    
    # Convert labels to a numpy array for Trainer compatibility
    tokenized_inputs["labels"] = np.array(labels)
    
    return tokenized_inputs


# Load the tokenizer and model
tokenizer = AutoTokenizer.from_pretrained("keshan/SinhalaBERTo")
model = AutoModelForSequenceClassification.from_pretrained("keshan/SinhalaBERTo", num_labels=2)  # Adjust num_labels if necessary

# Split the data into training and test sets
train_df, test_df = train_test_split(data, test_size=0.2, random_state=42)

# Convert pandas DataFrames to datasets Dataset
train_dataset = Dataset.from_pandas(train_df)
test_dataset = Dataset.from_pandas(test_df) 


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at keshan/SinhalaBERTo and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [66]:
train_dataset_business

Dataset({
    features: ['Comment ID', 'Comment', 'Business Sentiment', 'Content Sentiment', '__index_level_0__', 'input_ids', 'attention_mask', 'labels'],
    num_rows: 567
})

In [67]:
train_dataset_content

(Dataset({
     features: ['Comment ID', 'Comment', 'Business Sentiment', 'Content Sentiment', '__index_level_0__', 'input_ids', 'attention_mask', 'labels'],
     num_rows: 567
 }),)

In [68]:
test_dataset_content

(Dataset({
     features: ['Comment ID', 'Comment', 'Business Sentiment', 'Content Sentiment', '__index_level_0__', 'input_ids', 'attention_mask', 'labels'],
     num_rows: 142
 }),)

In [69]:
from transformers import Trainer, TrainingArguments

# Define the training arguments
training_args = TrainingArguments(
    output_dir="./results",          # output directory
    evaluation_strategy="epoch",     # evaluation strategy
    learning_rate=2e-5,              # learning rate
    per_device_train_batch_size=16,  # batch size for training
    per_device_eval_batch_size=16,   # batch size for evaluation
    num_train_epochs=3,              # number of training epochs
    weight_decay=0.01,               # strength of weight decay
)

In [70]:
# Define the compute_metrics function for evaluation
def compute_metrics(eval_pred):
    logits, labels = eval_pred
    preds = np.argmax(logits, axis=-1)
    # Assuming binary classification for simplicity
    accuracy = (preds == labels).mean()
    return {"accuracy": accuracy}

In [77]:
# from transformers import default_data_collator

# Initialize the Trainer for Content Sentiment
trainer_content = Trainer(
    model=model,                         # the model to be trained
    args=training_args,                  # training arguments
    train_dataset=train_dataset.map(preprocess_content, batched=True),   # training dataset with preprocessing
    eval_dataset=test_dataset.map(preprocess_content, batched=True),      # evaluation dataset with preprocessing
    compute_metrics=compute_metrics,     # evaluation metrics
)

# Fine-tune the model for Business Sentiment
trainer_content.train()

Map:   0%|          | 0/567 [00:00<?, ? examples/s]

Map:   0%|          | 0/142 [00:00<?, ? examples/s]

Epoch,Training Loss,Validation Loss,Accuracy
1,No log,0.544717,0.739437
2,No log,0.494225,0.732394
3,No log,0.459979,0.774648


TrainOutput(global_step=108, training_loss=0.47586423379403575, metrics={'train_runtime': 642.0593, 'train_samples_per_second': 2.649, 'train_steps_per_second': 0.168, 'total_flos': 56331761278464.0, 'train_loss': 0.47586423379403575, 'epoch': 3.0})

In [79]:
# Initialize the Trainer for Business Sentiment
trainer_business = Trainer(
    model=model,                         # the model to be trained
    args=training_args,                  # training arguments
    train_dataset=train_dataset.map(preprocess_business, batched=True),  # training dataset with preprocessing
    eval_dataset=test_dataset.map(preprocess_business, batched=True),     # evaluation dataset with preprocessing
    compute_metrics=compute_metrics,     # evaluation metrics
)

# Fine-tune the model for Business Sentiment
trainer_business.train()

Map:   0%|          | 0/567 [00:00<?, ? examples/s]

Map:   0%|          | 0/142 [00:00<?, ? examples/s]

Epoch,Training Loss,Validation Loss,Accuracy
1,No log,0.506763,0.781690
2,No log,0.530166,0.753521
3,No log,0.554204,0.802817


TrainOutput(global_step=108, training_loss=0.17344352934095594, metrics={'train_runtime': 647.9758, 'train_samples_per_second': 2.625, 'train_steps_per_second': 0.167, 'total_flos': 56331761278464.0, 'train_loss': 0.17344352934095594, 'epoch': 3.0})

In [81]:
# Save the model
trainer_content.save_model("./results_content")
trainer_business.save_model("./results_business")